internal/controller/meshfederation/meshfederation_controller.go

```diff
@@ -16,79 +16,238 @@ package meshfederation
 	})
 
-	if conditionsChanged {
+	if !reflect.DeepEqual(original.Status, meshFederation.Status) {
 		conditions := slices.Clone(meshFederation.Status.Conditions)
-		// TODO: patch and call only when actually conditionsChanged
+		// TODO: patch and call only when conditions were changed
 		_, errStatusUpdate := controller.RetryStatusUpdate(ctx, r.Client, meshFederation, func(saved *v1alpha1.MeshFederation) {
+			// TODO(design): do we want to keep all the services? seems convenient, but shouldn't we be concerned about the size?
+			saved.Status.ExportedServices = meshFederation.Status.ExportedServices
 			for _, condition := range conditions {
 				machinerymeta.SetStatusCondition(&saved.Status.Conditions, condition)
 			}
 		})
-		return ctrl.Result{}, errStatusUpdate
+
+		return reconcile.Result{}, errStatusUpdate
 	}
 
-	return ctrl.Result{}, nil
+	return reconcile.Result{}, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
+func (r *Reconciler) SetupWithManager(mgr manager.Manager) error {
+	return builder.ControllerManagedBy(mgr).
 		Named("mesh-federation-ctrl").
 		For(&v1alpha1.MeshFederation{}).
+		Owns(&v1alpha3.EnvoyFilter{}).
+		Owns(&v1beta1.PeerAuthentication{}).
+		Owns(&v1alpha3.Gateway{}).
+		Owns(&routev1.Route{}).
+		Watches(
+			// TODO(design): initial reconcile will trigger a lot of requests - one for each service. This can become expensive.
+			&corev1.Service{},
+			handler.EnqueueRequestsFromMapFunc(r.handleServiceToExport),
+			builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}, predicate.LabelChangedPredicate{})),
+		).
 		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, controller.FinalizerChanged())).
 		Complete(r)
 }
+
+func (r *Reconciler) subReconcile(ctx context.Context, meshFederation *v1alpha1.MeshFederation, exportedServices *corev1.ServiceList) (reconcile.Result, error) {
+	reconcilers := []controller.SubReconciler[*v1alpha1.MeshFederation]{
+		IngressGatewayReconciler{exportedServices: exportedServices}.Reconcile,
+		PeerAuth,
+	}
+
+	if meshFederation.Spec.IngressConfig.Type == "openshift-router" {
+		reconcilers = append(
+			reconcilers,
+			EnvoyFilter{exportedServices: exportedServices}.Reconcile,
+			RouteReconciler{exportedServices: exportedServices}.Reconcile,
+		)
+	}
+
+	var errs []error
+	var accResult reconcile.Result
+
+	for _, subreconciler := range reconcilers {
+		result, errSub := subreconciler(ctx, r.Client, meshFederation)
+		if errSub != nil {
+			errs = append(errs, errSub)
+		}
+
+		if result.Requeue {
+			accResult.Requeue = true
+		}
+
+		if result.RequeueAfter > accResult.RequeueAfter {
+			accResult.RequeueAfter = result.RequeueAfter
+		}
+	}
+
+	return accResult, errors.Join(errs...)
+}
+
+func (r *Reconciler) handleServiceToExport(ctx context.Context, object client.Object) []reconcile.Request {
```

Comment:
Nit: I think the name could be a little clearer. Something like this:

```suggestion
func (r *Reconciler) mapServiceToReconcileRequests(ctx context.Context, object client.Object) []reconcile.Request {
```